javalib/src/main/scala/java/util/Base64.scala

```diff
@@ -0,0 +1,486 @@
+        ByteBuffer.wrap(dst, 0, written)
+      } catch {
+        case e: IllegalArgumentException =>
+          buffer.position(start)
+          throw e
+      }
+    }
+
+    def wrap(is: InputStream): InputStream = new DecodingInputStream(is, table, ignoreInvalid)
+
+    // ------------------------------------------------------------------------
+    // PRIVATE
+    // ------------------------------------------------------------------------
+
+    private def doDecode(src: Wrapper[Byte], dst: Wrapper[Byte]): Int = {
+      val srcBuffer = new Wrapper[Int](new Array[Int](4))
+
+      @inline
+      def inputData(): Unit = {
+        srcBuffer.position = 0
+        var shift = 18
+        var i = 0
+        while(srcBuffer.hasRemaining) {
+          i |= (srcBuffer.get << shift)
+          shift -= 6
+        }
+        if(shift == 12) throw new IllegalArgumentException("Last unit does not have enough valid bits")
+        if(shift <= 6) dst.put((i >> 16).toByte)
+        if(shift <= 0) dst.put((i >> 8).toByte)
+        if(shift <= -6) dst.put(i.toByte)
+        srcBuffer.clear()
+      }
+
+      @inline
+      def iterate(): Unit = {
+        // scalastyle:off return
+        while (src.hasRemaining) {
+          if (srcBuffer.hasRemaining) {
+            val int = src.get & 0xff
+            table(int) match {
+              case -2 => return
+              case -1 if ignoreInvalid =>
+              case -1 => throw new IllegalArgumentException("Illegal base64 character " + Integer.toString(int, 16))
+              case i  => srcBuffer.put(i)
+            }
+          } else inputData()
+        }
+        // scalastyle:on return
+      }
+      iterate()
+      // end or padding
+      srcBuffer.flip()
+      inputData()
+      while(src.hasRemaining) {
+        val int = src.get & 0xff
+        val value = table(int)
+        if (value != -2 && (!ignoreInvalid || value > 0))
+          throw new IllegalArgumentException(s"Input byte array has incorrect ending byte at $int")
+      }
+      dst.position
+    }
+
+    private def dstRequiredLength(src: Array[Byte], offset: Int, length: Int): Int = {
+      // scalastyle:off return
+      if (length == 0) return 0
+      // scalastyle:on return
+
+      var validBytes = 0
+      var padding = 0
+      if (ignoreInvalid) {
+        for (i <- offset until length) {
+          if (table(src(i) & 0xff) >= 0)
+            validBytes += 1
+        }
+      } else {
+        validBytes = length-offset
+        /* We check the end for padding and compute the length from there. This is ok, if the rest contains garbage
+        we'll have written something before throwing but the spec says "If the input byte array is not in valid Base64
+        encoding scheme then some bytes may have been written to the output byte array before IllegalArgumentException
+        is thrown."*/
+        if (src(length-1) == '=') {
+          padding += 1
+          if (src(length-2) == '=')
+          padding += 1
```

Comment:
Also, I think this can cause an `IndexOutOfBoundsException` if the input buffer contains exactly `"="`. If you eliminate the `if (length == 0)` at the beginning of the method, it can also crash when the input buffer is empty. You should guard against those scenarios, e.g.,
```scala
if (length >= 1 && src(length - 1) == '=')
```
(there should be tests for those scenarios)